File: src/main/java/Piruru/actions/MillAction.java
package Piruru.actions;

import com.megacrit.cardcrawl.actions.AbstractGameAction;
import com.megacrit.cardcrawl.cards.AbstractCard;
import com.megacrit.cardcrawl.cards.blue.Consume;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.vfx.cardManip.ShowCardAndAddToDiscardEffect;

import java.util.function.Consumer;


public class MillAction extends AbstractGameAction {
    private int amount;
    private Consumer<AbstractCard> callback = null;

    public MillAction(int amount) {
        this.amount = amount;
    }

    public MillAction(int amount, Consumer<AbstractCard> callback) {
        this(amount);
        this.callback = callback;
    }

    @Override
    public void update() {
        if (AbstractDungeon.player.drawPile.size() == 0) {
            isDone = true;
            return;
        }
        AbstractCard c = AbstractDungeon.player.drawPile.getTopCard();
        AbstractDungeon.player.drawPile.removeCard(c);
        AbstractDungeon.player.discardPile.removeCard(c);
        AbstractDungeon.effectsQueue.add(new ShowCardAndAddToDiscardEffect(c));
        amount--;
        if (amount > 0) {
            if (callback != null) {
                callback.accept(c);
                AbstractDungeon.actionManager.addToBottom(new MillAction(amount, callback));
                isDone = true;
                return;
            }
            AbstractDungeon.actionManager.addToBottom(new MillAction(amount));
        }
        isDone = true;
    }
}

File: src/main/kotlin/Cirno/actions/MillAction.kt
package Cirno.actions

import com.megacrit.cardcrawl.actions.AbstractGameAction
import com.megacrit.cardcrawl.cards.AbstractCard
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.vfx.cardManip.ShowCardAndAddToDiscardEffect


class MillAction(private var millAmount: Int) : AbstractGameAction() {
    private var callback: ((AbstractCard) -> Unit)? = null

    constructor(amount: Int, callback: ((AbstractCard) -> Unit)?) : this(amount) {
        this.callback = callback
    }

    override fun update() {
        if (AbstractDungeon.player.drawPile.size() == 0) {
            isDone = true
            return
        }
        val c = AbstractDungeon.player.drawPile.topCard
        AbstractDungeon.player.drawPile.removeCard(c)
        AbstractDungeon.player.discardPile.removeCard(c)
        AbstractDungeon.effectsQueue.add(ShowCardAndAddToDiscardEffect(c))
        millAmount--
        if (millAmount > 0) {
            if (callback != null) {
                callback!!(c)
                AbstractDungeon.actionManager.addToBottom(MillAction(millAmount, callback))
                isDone = true
                return
            }
            AbstractDungeon.actionManager.addToBottom(MillAction(millAmount))
        }
        isDone = true
    }
}
